src/backends/cg.rs

```diff
@@ -29,65 +34,231 @@ impl AsRef<[u8]> for Buffer {
+            _object: Option<&AnyObject>,
+            change: Option<&NSDictionary<NSKeyValueChangeKey, AnyObject>>,
+            _context: *mut c_void,
+        ) {
+            self.update(key_path, change);
+        }
+    }
+);
+
+// SAFETY: The `CALayer` that the observer contains is thread safe.
+unsafe impl Send for Observer {}
+unsafe impl Sync for Observer {}
+
+impl Observer {
+    fn new(layer: &CALayer) -> Retained<Self> {
+        let this = Self::alloc().set_ivars(layer.retain());
+        unsafe { msg_send_id![super(this), init] }
+    }
+
+    fn update(
+        &self,
+        key_path: Option<&NSString>,
+        change: Option<&NSDictionary<NSKeyValueChangeKey, AnyObject>>,
+    ) {
+        let layer = self.ivars();
+
+        let change =
+            change.expect("requested a change dictionary in `addObserver`, but none was provided");
+        let new = change
+            .get(unsafe { NSKeyValueChangeNewKey })
+            .expect("requested change dictionary did not contain `NSKeyValueChangeNewKey`");
+
+        // NOTE: Setting these values usually causes a quarter second animation to occur, which is
+        // undesirable.
+        //
+        // However, since we're setting them inside an observer, there already is a transaction
+        // ongoing, and as such we don't need to wrap this in a `CATransaction` ourselves.
+
+        if key_path == Some(ns_string!("contentsScale")) {
+            let new = unsafe { &*(new as *const AnyObject as *const NSNumber) };
+            let scale_factor = new.as_cgfloat();
+
+            // Set the scale factor of the layer to match the root layer when it changes (e.g. if
+            // moved to a different monitor, or monitor settings changed).
+            layer.setContentsScale(scale_factor);
+        } else if key_path == Some(ns_string!("bounds")) {
+            let new = unsafe { &*(new as *const AnyObject as *const NSValue) };
+            let bounds = new.get_rect().expect("new bounds value was not CGRect");
+
+            // Set `bounds` and `position` so that the new layer is inside the superlayer.
+            //
+            // This differs from just setting the `bounds`, as it also takes into account any
+            // translation that the superlayer may have that we'd want to preserve.
+            layer.setFrame(bounds);
+        } else {
+            panic!("unknown observed keypath {key_path:?}");
+        }
+    }
+}
+
 pub struct CGImpl<D, W> {
-    layer: MainThreadBound<Id<CALayer>>,
-    window: MainThreadBound<Id<NSWindow>>,
+    /// Our layer.
+    layer: SendCALayer,
+    /// The layer that our layer was created from.
+    ///
+    /// Can also be retrieved from `layer.superlayer()`.
+    root_layer: SendCALayer,
+    observer: Retained<Observer>,
     color_space: SendCGColorSpace,
-    size: Option<(NonZeroU32, NonZeroU32)>,
+    /// The width of the underlying buffer.
+    width: u32,
+    /// The height of the underlying buffer.
+    height: u32,
```

Comment:
It seems that the internal API is mostly using `usize`, it might be good to store them as `usize` as well and do the conversion once instead of everywhere else.